data_iterator.go

```diff
@@ -185,22 +190,28 @@ func (d *DataIterator) Run() {
 			defer wg.Done()
 
 			for {
-				table, ok := <-tablesQueue
+				var table *schema.Table
+				var ok bool
+				select {
+				case table, ok = <-tablesQueue:
+				case <-ctx.Done():
+					logrus.WithField("tag", "table_iterator").Info("cancellation received, terminating goroutine abruptly")
+				}
 				if !ok {
 					break
 				}
 
 				logger := d.logger.WithField("table", table.String())
 
 				cursor := d.CursorConfig.NewCursor(table, d.CurrentState.TargetPrimaryKeys()[table.String()])
-				err := cursor.Each(func(batch *RowBatch) error {
+				err := cursor.Each(ctx, func(batch *RowBatch) error {
 					metrics.Count("RowEvent", int64(batch.Size()), []MetricTag{
 						MetricTag{"table", table.Name},
 						MetricTag{"source", "table"},
 					}, 1.0)
 
 					for _, listener := range d.batchListeners {
-						err := listener(batch)
+						err := listener(ctx, batch)
```

Comment:
Again, I'm not sure why the listener needs a context. There's no rush to terminate _immediately_.

We want to finish up writing our batch, record the last successful PK, and _then_ check the termination signal.

i.e. the only places that the DataIterator should check for interruption are:
(1) After a batch has been written and its bookkeeping has been updated
(2) The producer of the `tablesQueue` channel